ipa-core/src/protocol/ipa_prf/malicious_security/lagrange.rs

```diff
@@ -0,0 +1,329 @@
+                );
+            });
+        });
+    }
+    // compute denominators:
+    denominators
+        .iter_mut()
+        .zip(evaluation_points.iter().enumerate())
+        .for_each(|(d, i)| {
+            evaluation_points
+                .iter()
+                .enumerate()
+                .filter(|&j| j.0 != i.0)
+                .for_each(|j| *d *= *i.1 - *j.1);
+            *d = d.invert();
+        });
+}
+
+/// computes the lagrange base polynomials, see `https://en.wikipedia.org/wiki/Lagrange_polynomial`
+/// for correctness, the denominators need to be consistent with the `evaluation_points`
+/// overwrites `base`
+///
+/// # Panics
+/// Does not panic
+pub fn compute_lagrange_base<F>(points: &[F], e_p: &[F], ds: &[F], base: &mut Vec<Vec<F>>)
+where
+    F: Field,
+{
+    debug_assert_eq!(e_p.len(), ds.len());
+    #[cfg(debug_assertions)]
+    {
+        for base in base.iter() {
+            assert_eq!(e_p.len(), base.len());
+        }
+    }
+    debug_assert_eq!(points.len(), base.len());
+
+    // for each point, compute base polynomials
+    points
+        .iter()
+        .zip(base.iter_mut())
+        .for_each(|(point, base)| {
+            // set up the denominators
+            base.copy_from_slice(ds);
+            // compute the nominator
+            base.iter_mut().enumerate().for_each(|b| {
+                e_p.iter()
+                    .enumerate()
+                    .filter(|&j| j.0 != b.0)
+                    .for_each(|j| {
+                        *b.1 *= *point - *j.1;
+                    });
+            });
+        });
+}
+
+/// precomputed lagrange evaluation
+/// evaluates polynomial using base polynomials
+/// the base polynomials are tied to specific evaluation points
+/// multiplies `result` with input `result`, i.e. use input `F::ONE`
+pub fn lagrange_evaluation_precomputed<F>(y: &[F], base: &Vec<Vec<F>>, result: &mut [F])
+where
+    F: Field,
+{
+    debug_assert_eq!(base.len(), result.len());
+    #[cfg(debug_assertions)]
+    {
+        for base in base {
+            debug_assert_eq!(y.len(), base.len());
+        }
+    }
+    result
+        .iter_mut()
+        .zip(base.iter())
+        .for_each(|(result, base)| {
+            *result *= base
+                .iter()
+                .zip(y.iter())
+                .fold(F::ZERO, |acc, (base, y)| acc + *base * *y);
+        });
+}
+
+/// lagrange evaluation
+/// given points `(x_0,y_0),...,(x_n,y_n)` on a polynomial `p` and points `p_0,...,p_m`
+/// the function computes `p` evaluated on points `p_0,...,p_m`
+///
+/// It follows the lagrange method to compute the points `https://en.wikipedia.org/wiki/Lagrange_polynomial`
+/// further, rather than outputting `p(p_k)`, we set `result_k=result_k*p(p_k)`
+/// which fits better to how we use `p(p_k)`
+///
+/// # Panics
+/// Panics when there are identical evaluation points, i.e. division by zero
+pub fn lagrange_evaluation<F>(x: &[F], y: &[F], p: &[F], result: &mut [F])
```

Comment:
I think we should just have one of the two approaches, presumably the precomputed version. Otherwise we have two independent implementations to keep in sync and maintain.